File: src/git.mjs
/**
 * Git facts for the context line.
 *
 * One `git status --porcelain=v1 -b` call covers branch, upstream divergence
 * and working-tree counts, so the statusline pays for a single process per
 * render instead of three.
 */

import { execFileSync } from "node:child_process";

const GIT_TIMEOUT_MS = 400;

function git(args, cwd) {
  return execFileSync("git", args, {
    cwd,
    encoding: "utf8",
    timeout: GIT_TIMEOUT_MS,
    stdio: ["ignore", "pipe", "ignore"],
    maxBuffer: 2_000_000,
  });
}

/**
 * @returns {{repo: string|null, branch: string|null, staged: number,
 *   modified: number, untracked: number, ahead: number, behind: number}|null}
 *   or null when `cwd` is not inside a work tree.
 */
export function readGit(cwd) {
  let raw;
  try {
    raw = git(["status", "--porcelain=v1", "-b", "--untracked-files=normal"], cwd);
  } catch {
    return null; // not a repo, git missing, or slower than the timeout
  }

  const result = {
    repo: null,
    branch: null,
    staged: 0,
    modified: 0,
    untracked: 0,
    ahead: 0,
    behind: 0,
  };

  for (const line of raw.split("\n")) {
    if (line === "") continue;
    if (line.startsWith("## ")) {
      // Header shapes: "main...origin/main [ahead 2]", "HEAD (no branch)",
      // and "No commits yet on main" for a freshly initialised repo.
      const header = line.slice(3).replace(/^No commits yet on /, "");
      const branchPart = header.split(/\.\.\.|\s+\[/)[0];
      result.branch = branchPart.startsWith("HEAD (no branch)") ? "detached" : branchPart;
      const ahead = header.match(/ahead (\d+)/);
      const behind = header.match(/behind (\d+)/);
      if (ahead) result.ahead = Number(ahead[1]);
      if (behind) result.behind = Number(behind[1]);
      continue;
    }
    if (line.startsWith("??")) {
      result.untracked += 1;
      continue;
    }
    // XY status: X is the index (staged), Y the work tree (unstaged).
    if (line[0] !== " " && line[0] !== "?") result.staged += 1;
    if (line[1] !== " " && line[1] !== "?") result.modified += 1;
  }

  try {
    const top = git(["rev-parse", "--show-toplevel"], cwd).trim();
    if (top) result.repo = top.split("/").pop() ?? null;
  } catch {
    /* branch alone is enough for the line */
  }

  return result;
}

File: git.go
package main

import (
	"context"
	"io"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const gitTimeout = 400 * time.Millisecond

var (
	aheadPattern       = regexp.MustCompile(`ahead (\d+)`)
	behindPattern      = regexp.MustCompile(`behind (\d+)`)
	branchSplitPattern = regexp.MustCompile(`\.\.\.|\s+\[`)
)

// GitInfo is what the context line shows about the working tree.
type GitInfo struct {
	Repo      string // basename of the top-level dir, "" when unknown
	Branch    string // "detached" on a detached HEAD
	Staged    int
	Modified  int
	Untracked int
	Ahead     int
	Behind    int
}

// readGit reads the working tree state from one porcelain status call, or nil
// when cwd is not a work tree, git is missing, or git is slower than gitTimeout
// (a statusline cannot wait on a huge repo). The repo name costs a second git
// process, so it is only read when wantRepo is set.
//
// Header shapes: "## main...origin/main [ahead 2, behind 1]", "## HEAD (no
// branch)" (Branch becomes "detached"), "## No commits yet on main" (the prefix
// is dropped). The branch is the header up to "..." or " [". Status lines: "??"
// is untracked; otherwise a non-space, non-"?" X column counts as staged and a
// non-space, non-"?" Y column as modified.
func readGit(cwd string, wantRepo bool) *GitInfo {
	run := func(args ...string) (string, bool) {
		ctx, cancel := context.WithTimeout(context.Background(), gitTimeout)
		defer cancel()
		cmd := exec.CommandContext(ctx, "git", args...)
		cmd.Dir = cwd
		cmd.Stderr = io.Discard
		out, err := cmd.Output()
		if err != nil {
			return "", false
		}
		return string(out), true
	}

	raw, ok := run("status", "--porcelain=v1", "-b", "--untracked-files=normal")
	if !ok {
		return nil
	}
	result := &GitInfo{}
	for _, line := range strings.Split(raw, "\n") {
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "## ") {
			header := strings.TrimPrefix(line[3:], "No commits yet on ")
			branch := branchSplitPattern.Split(header, 2)[0]
			if strings.HasPrefix(branch, "HEAD (no branch)") {
				result.Branch = "detached"
			} else {
				result.Branch = branch
			}
			if match := aheadPattern.FindStringSubmatch(header); match != nil {
				result.Ahead, _ = strconv.Atoi(match[1])
			}
			if match := behindPattern.FindStringSubmatch(header); match != nil {
				result.Behind, _ = strconv.Atoi(match[1])
			}
			continue
		}
		if strings.HasPrefix(line, "??") {
			result.Untracked++
			continue
		}
		if line[0] != ' ' && line[0] != '?' {
			result.Staged++
		}
		if len(line) > 1 && line[1] != ' ' && line[1] != '?' {
			result.Modified++
		}
	}

	if !wantRepo {
		return result
	}
	if top, ok := run("rev-parse", "--show-toplevel"); ok {
		top = strings.TrimSpace(top)
		result.Repo = top[strings.LastIndex(top, "/")+1:]
	}
	return result
}
